File: array_of_products/array_of_products.ts
function arrOfProducts(array: number[]): number[] {
    const products: number[] = new Array(array.length).fill(1);
    let product = 1;

    for (let i = array.length - 1; i >= 0; i--) {
        products[i] = product;
        product *= array[i];
    }

    product = 1;

    for (let i = 0; i < array.length; i++) {
        products[i] *= product;
        product *= array[i];
    }

    return products;
}

console.log(arrOfProducts([5, 1, 4, 2]));

File: array_of_products/array_of_products.py
# O(n) time | O(n) space - Where n is the length of the input array
def array_of_products(array):
  products = [1 for _ in range(len(array))]
  product = 1
  for i in reversed(range(len(array))):
    products[i] = product
    product *= array[i]
  product = 1
  for i in range(len(array)):
    products[i] = product * products[i]
    product *= array[i]
  return products

print(array_of_products([5, 1, 4, 2]))
